OSDOCS Speeding Up Pulling Container Images/CRI-O Additional Storage Support (#110809)
openshift/openshift-docs: modules/nodes-nodes-additional-crio-storage-about.adoc

// Module included in the following assemblies:
//

> // * nodes/pods/nodes-pods-image-volume.adoc

Reviewer: Is this path correct? Should it be:

```suggestion
// * nodes/nodes/nodes-nodes-additional-crio-storage.adoc
```

:_mod-docs-content-type: CONCEPT
[id="nodes-nodes-additional-crio-storage-about_{context}"]
= About additional storage locations for CRI-O

[role="_abstract"]

> You can configure additional storage locations for the CRI-O container engine that give you control over where CRI-O stores and retrieves OCI artifacts, complete container images, and container image layers. Using dedicated storage locations for these CRI-O objects can reduce application startup time and make your applications run more efficiently through dedicated solid-state drive (SSD) storage, shared image caches, or lazy pulling.

Reviewer 2: This also needs to come in under 300 (currently 439)

By default, CRI-O stores all container data under a single root directory, `/var/lib/containers/storage`. This works well for typical workloads, but creates problems for artificial intelligence and machine learning (AI/ML) workloads and other scenarios that use large images or artifacts. 

> For example, large OCI artifacts, such as machine learning models, are stored on the default location, consuming space and preventing the use of faster dedicated storage. By configuring the `additionalArtifactStores` parameter, you can store large AI/ML models on high-performance storage (SSD) separate from the root file system space storage. As a result, your workloads can experience faster start times and your clusters can use storage more efficiently.

Reviewer: "root file system space storage" is awkward.

```suggestion
For example, large OCI artifacts, such as machine learning models, are stored on the default location, consuming space and preventing the use of faster dedicated storage. By configuring the `additionalArtifactStores` parameter, you can store large AI/ML models on high-performance storage (SSD) separate from the root filesystem. As a result, your workloads can experience faster start times and your clusters can use storage more efficiently.
```

> Also, you could use the `additionalImageStores` parameter to mount an NFS share with pre-populated images across all worker nodes. Nodes read from the shared cache instead of pulling from an external registry. This is useful in a disconnected environments or when many nodes run the same workloads.

Reviewer: Grammar: "in a disconnected environments"

```suggestion
Also, you could use the `additionalImageStores` parameter to mount an NFS share with pre-populated images across all worker nodes. Nodes read from the shared cache instead of pulling from an external registry. This is useful in disconnected environments or when many nodes run the same workloads.
```

> With the `additionalLayerStores` parameter, you can enable lazy-image pulling through a third-party storage plugin, such as stargz-store. With lazy pulling, containers start after downloading only the required file chunks. The remaining data is fetched during runtime.

Reviewer: "lazy-image pulling" reads as "pulling of lazy images." Use "lazy pulling" for consistency.

```suggestion
With the `additionalLayerStores` parameter, you can enable lazy pulling through a third-party storage plugin, such as stargz-store. With lazy pulling, containers start after downloading only the required file chunks. The remaining data is fetched during runtime.
```

> With the `additionalLayerStores` parameter, you could enable lazy pulling through a third-party storage plugin, such as stargz-store. With lazy pulling, containers start after downloading only the required file chunks. The remaining data is fetched during runtime.

Reviewer 3: ```suggestion
For example, large OCI artifacts, such as machine learning models, are stored on the default location, consuming space and preventing the use of faster dedicated storage. By configuring the `additionalArtifactStores` field, you can store large AI/ML models on high-performance storage (SSD) separate from the root file system. As a result, your workloads can experience faster start times and your clusters can use storage more efficiently.

Also, you could use the `additionalImageStores` field to mount an NFS share with pre-populated images across all worker nodes. Nodes read from the shared cache instead of pulling from an external registry. This is useful in disconnected environments or when many nodes run the same workloads.

With the `additionalLayerStores` field, you could enable lazy pulling through a third-party storage plugin, such as stargz-store. With lazy pulling, containers start after downloading only the required file chunks. The remaining data is fetched during runtime.
```

After you configure any of these new storage locations, the Machine Config Operator (MCO) reboots the affected nodes with the new configuration. After the reboot, CRI-O begins resolving storage from the additional locations.

Additional storage for OCI artifacts::

> Use the `additionalArtifactStores` field in a container runtime config to specify read-only locations where CRI-O resolves OCI artifacts, such as machine learning models pulled as OCI volume images. CRI-O checks these locations in order before falling back to the default storage location.

Reviewer: CRI-O expects an `artifacts/` subdirectory within each configured path (see cri-o/cri-o#9702). Without this, the store appears empty.

```suggestion
Use the `additionalArtifactStores` field in a container runtime config to specify read-only locations where CRI-O resolves OCI artifacts, such as machine learning models pulled as OCI volume images. CRI-O checks these locations in order before falling back to the default storage location. CRI-O expects an `artifacts/` subdirectory within each configured path. For example, if the path is `/mnt/ssd-artifacts`, place the artifacts in `/mnt/ssd-artifacts/artifacts/`.
```

Author: I don't think this is in your `docs-draft-additional-storage-config.md` file. Thanks!

> +

Reviewer 3: Can we add note about mounting storage 
```suggestion
[NOTE]                                                                                                                                                                                                            
====                                                                                                                                                                                                              
The specified paths must already exist and be accessible on the nodes. For network storage, ensure the paths are mounted before applying the configuration.
====
```

Author: @BhargaviGudi The first sentence is addressed in the [prerequisites for the procedure](https://github.com/openshift/openshift-docs/pull/110809/changes#diff-224a349fbf80430edcdceb47812447b7c9615937d4db597e85a788f70a70bb2aR28) (where I added your second setnence). It seems to me that it makes more sense to put these items there, where users need this level of detail. It might be too much information here. WDYT?

> Use the `additionalArtifactStores` field in a container runtime config to specify read-only locations where CRI-O resolves OCI artifacts, such as machine learning models pulled as OCI volume images. CRI-O checks these locations in order before falling back to the default storage location. CRI-O requires an existing, pre-populated `artifacts/` subdirectory within each configured path. For example, if the path is `/mnt/ssd-artifacts`, place the artifacts in the `/mnt/ssd-artifacts/artifacts/` directory.

Reviewer 2: ```suggestion
Use the `additionalArtifactStores` field in a container runtime config to specify read-only locations where CRI-O resolves OCI artifacts, such as machine learning models pulled as OCI volume images. CRI-O checks these locations in order before falling back to the default storage location. CRI-O requires an existing, prepopulated `artifacts/` subdirectory within each configured path. For example, if the path is `/mnt/ssd-artifacts`, place the artifacts in the `/mnt/ssd-artifacts/artifacts/` directory.
```

[ISG hates hyphens](https://www.ibm.com/docs/en/ibm-style?topic=grammar-prefixes#hyphens)

The following example container runtime config configures storage for OCI artifacts.
+
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: ContainerRuntimeConfig
metadata:
  name: ssd-artifact-stores
spec:
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ""
  containerRuntimeConfig:
    additionalArtifactStores:
      - path: /mnt/ssd-artifacts
      - path: /mnt/nfs-shared-artifacts
----
+
When you create the container runtime config, the Machine Config Operator (MCO) writes the configuration into the `/etc/crio/crio.conf.d/01-ctrcfg-additionalArtifactStores` on the target nodes.

> Additional storage for OCI artifacts::

Reviewer: Copy-paste error: this describes `additionalImageStores` but the term says "OCI artifacts" again.

```suggestion
Additional storage for container images::
```

Use the `additionalImageStores` field to specify read-only container image caches on shared or high-performance storage. When CRI-O needs an image, it checks the additional image stores first. If the image exists there, no registry pull happens.

> +

Reviewer 3: Same as https://github.com/openshift/openshift-docs/pull/110809/changes#r3186664497

Author: Same as https://github.com/openshift/openshift-docs/pull/110809#discussion_r3189631985

The following example container runtime config configures storage for container images.
+
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: ContainerRuntimeConfig
metadata:
  name: shared-image-cache
spec:
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ""
  containerRuntimeConfig:
    additionalImageStores:
      - path: /mnt/nfs-image-cache
      - path: /mnt/ssd-images
----
+
When you create the container runtime config, the Machine Config Operator (MCO) writes the configuration into the `/etc/containers/storage.conf` on the target nodes.

Additional container image layers::

> Use the `additionalLayerStores` field to enable lazy-image pulling by a third-party storage plugin. 

Reviewer: Same hyphenation issue as line 18.

```suggestion
Use the `additionalLayerStores` field to enable lazy pulling through a third-party storage plugin. 
```

> +

Reviewer 3: Can we add note regarding storage plugin
```suggestion
[IMPORTANT]                                                                                                                                                                                                       
====                                                                                                                                                                                                              
Before configuring layer stores:
- Install a compatible storage plugin (such as Nydus Storage Plugin or Stargz Store) on all nodes. The plugin must be running before CRI-O can use layer stores. See "Additional resources" for installation guides.                                                                                                                                                                                                           
- Convert container images to a lazy-pull format (eStargz or Nydus). Standard OCI images will fall back to regular pulls. See "Additional resources" for image conversion tools.
- Verify your container registry supports HTTP range requests.                                                                                                                                                    
- The Machine Config Operator automatically appends the `:ref` suffix to layer store paths in `/etc/containers/storage.conf`. Do not include this suffix in your ContainerRuntimeConfig.                          
==== 
```
cc: @saschagrunert

Author: @BhargaviGudi Same as https://github.com/openshift/openshift-docs/pull/110809#discussion_r3189631985.
See https://github.com/openshift/openshift-docs/pull/110809/changes#diff-224a349fbf80430edcdceb47812447b7c9615937d4db597e85a788f70a70bb2aR32-R38 and https://github.com/openshift/openshift-docs/pull/110809/changes#diff-224a349fbf80430edcdceb47812447b7c9615937d4db597e85a788f70a70bb2aR84

Note that CRI-O falls back to a standard image pull in the following cases:
+
--
* The registry does not support HTTP range requests.
* The image is in standard OCI format, not a lazy-pull-compatible format such as eStargz or Nydus.
* The storage plugin is not running.
--
+
The following example container runtime config configures container image layers for lazy pulling.
+
[source,yaml]
----
apiVersion: machineconfiguration.openshift.io/v1
kind: ContainerRuntimeConfig
metadata:
  name: lazy-pulling
spec:
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ""
  containerRuntimeConfig:
    additionalLayerStores:
      - path: /var/lib/stargz-store
----
+

> When you create the container runtime config, the Machine Config Operator (MCO) writes the configuration into the `/etc/containers/storage.conf` on the target nodes.

Reviewer 3: ```suggestion
When you create the container runtime config, the Machine Config Operator (MCO) writes the configuration into the `/etc/containers/storage.conf` file on the target nodes.
```

> When you create the container runtime config, the Machine Config Operator (MCO) writes the configuration to the `/etc/containers/storage.conf` file on the target nodes.

Reviewer 2: This feels like a separate reference module about "additional storage location fields" or something

Author: These are examples of how you create the additional storage locations that are described earlier in the module. If I separate these examples, I would have to explain what additional storage is again. Seems redundant to me.
